common/windows-app-bootstrap.cpp

```diff
@@ -10,11 +10,141 @@
+    else if (code == EXCEPTION_NONCONTINUABLE_EXCEPTION) return "Noncontinuable exception occured!";
+    else if (code == EXCEPTION_PRIV_INSTRUCTION) return "Error due to invalid call to priviledged instruction!";
+    else if (code == EXCEPTION_STACK_OVERFLOW) return "Stack overflow error!";
+    else if (code == VcppException(ERROR_SEVERITY_ERROR, ERROR_MOD_NOT_FOUND)) {
+        auto details = (DelayLoadInfo*)ep->ExceptionRecord->ExceptionInformation[0];
+        std::string dll_name = details->szDll;
+        std::string fname = details->dlp.szProcName;
+        return rs2::to_string() << "Could not find " << dll_name << " library required for " << fname << ".\nMake sure all program dependencies are reachable or download standalone version of the App from our GitHub";
+    }
+    else return rs2::to_string() << "Unknown error (" << code << ")!";
+}
+
+// Show custom error message box with OK and Report options
+// Report will prompt new GitHub issue
+void report_error(std::string error)
+{
+    std::wstring ws(error.begin(), error.end());
+    SetWindowsHookEx(WH_CBT, &cbtproc, 0, GetCurrentThreadId());
+    auto button = MessageBox(NULL, ws.c_str(), L"Something went wrong...", MB_ICONERROR | MB_OKCANCEL);
+    if (button == IDCANCEL)
+    {
+        std::stringstream ss;
+        rs2_error* e = nullptr;
+
+        ss << "| | |\n";
+        ss << "|---|---|\n";
+        ss << "|**librealsense**|" << rs2::api_version_to_string(rs2_get_api_version(&e)) << (rs2::is_debug() ? " DEBUG" : " RELEASE") << "|\n";
+        ss << "|**OS**|" << rs2::get_os_name() << "|\n\n";
+        ss << "Intel RealSense Viewer / Depth Quality Tool has crashed with the following error message:\n";
+        ss << "```\n";
+        ss << error;
+        ss << "\n```";
+
+        std::string link = "https://github.com/IntelRealSense/librealsense/issues/new?body=" + rs2::url_encode(ss.str());
+        rs2::open_url(link.c_str());
+    }
+}
+
+// Global crash handler will be triggered when exception escapes from thread
+LONG WINAPI CrashHandler(EXCEPTION_POINTERS* ep)
+{
+    auto code = ep->ExceptionRecord->ExceptionCode;
+
+    if (should_intercept(code))
+    {
+        std::string error = "Unhandled exception escaping from a worker thread!\nError type: ";
+        error += exception_code_to_string(code, ep);
+        report_error(error);
+    }
+
+    return EXCEPTION_CONTINUE_SEARCH;
+}
+
+// Actual main - implemented in the cross-platform section of the App
 int main(int argv, const char** argc);
 
+int filter(unsigned int code, struct _EXCEPTION_POINTERS *ep)
+{
+    if (should_intercept(code))
+    {
+        auto error = exception_code_to_string(code, ep);
+        std::cerr << "Program terminated due to an unrecoverable SEH exception:\n" << error;
+        return EXCEPTION_EXECUTE_HANDLER;
+    }
+    else return EXCEPTION_CONTINUE_SEARCH;
+}
+
+// Wrapper around main. Must be in a function without destructors (that does not require stack unwinding)
```

Comment:
Doesn't WinMain below have dtors? vector, stringstream?

Reply:
The rule that you can have structured exception handler or regular C++ exception handler, for each function. If you have objects with destructors you can't have SEH in the same function, but you can in a separate function (hence the extra function)

Reply:
Then the comment is confusing :)